src/tools/w3dviewer/datatreeview.cpp

```diff
@@ -0,0 +1,680 @@
+    m_imageListIDs[ICON_SOUND] = -1;
+}
+
+BOOL CDataTreeView::PreCreateWindow(CREATESTRUCT &cs)
+{
+    cs.style |= (TVS_SHOWSELALWAYS | TVS_LINESATROOT | TVS_HASLINES | TVS_HASBUTTONS);
+    return CTreeView::PreCreateWindow(cs);
+}
+
+void CDataTreeView::OnInitialUpdate()
+{
+    CTreeView::OnInitialUpdate();
+}
+
+int CDataTreeView::OnCreate(LPCREATESTRUCT lpCreateStruct)
+{
+    int result = CTreeView::OnCreate(lpCreateStruct);
+
+    if (result != -1) {
+        CImageList il;
+        il.Create(16, 18, ILC_MASK, 5, 10);
+        m_imageListIDs[ICON_ANIMATION] = il.Replace(-1, LoadIcon(AfxGetResourceHandle(), MAKEINTRESOURCE(IDI_ANIMATION)));
+        m_imageListIDs[ICON_ANIMSPEED] = il.Replace(-1, LoadIcon(AfxGetResourceHandle(), MAKEINTRESOURCE(IDI_ANIMSPEED)));
+        m_imageListIDs[ICON_TEXTURE] = il.Replace(-1, LoadIcon(AfxGetResourceHandle(), MAKEINTRESOURCE(IDI_TEXTURE)));
+        m_imageListIDs[ICON_MESH] = il.Replace(-1, LoadIcon(AfxGetResourceHandle(), MAKEINTRESOURCE(IDI_MESH)));
+        m_imageListIDs[ICON_MATERIAL] = il.Replace(-1, LoadIcon(AfxGetResourceHandle(), MAKEINTRESOURCE(IDI_MATERIAL)));
+        m_imageListIDs[ICON_HLOD] = il.Replace(-1, LoadIcon(AfxGetResourceHandle(), MAKEINTRESOURCE(IDI_HEIRARCHY)));
+        m_imageListIDs[ICON_AGGREGATE] = il.Replace(-1, LoadIcon(AfxGetResourceHandle(), MAKEINTRESOURCE(IDI_HEIRARCHY)));
+        m_imageListIDs[ICON_EMITTER] = il.Replace(-1, LoadIcon(AfxGetResourceHandle(), MAKEINTRESOURCE(IDI_HEIRARCHY)));
+        m_imageListIDs[ICON_HIERARCHY] = il.Replace(-1, LoadIcon(AfxGetResourceHandle(), MAKEINTRESOURCE(IDI_HEIRARCHY)));
+        m_imageListIDs[ICON_PRIMITIVE] = il.Replace(-1, LoadIcon(AfxGetResourceHandle(), MAKEINTRESOURCE(IDI_PRIMITIVE)));
+        m_imageListIDs[ICON_SOUND] = il.Replace(-1, LoadIcon(AfxGetResourceHandle(), MAKEINTRESOURCE(IDI_SOUND)));
+        GetTreeCtrl().SetImageList(&il, 0);
+        il.Detach();
+        AddCategories();
+        return 0;
+    }
+
+    return result;
+}
+
+void CDataTreeView::OnSelectTree(NMHDR *pNMHDR, LRESULT *pResult)
+{
+    NMTREEVIEW *nm = (NMTREEVIEW *)pNMHDR;
+    Select(nm->itemNew.hItem);
+    *pResult = 0;
+}
+
+void CDataTreeView::OnDeleteItem(NMHDR *pNMHDR, LRESULT *pResult)
+{
+    NMTREEVIEW *nm = (NMTREEVIEW *)pNMHDR;
+    AssetInfoClass *info = (AssetInfoClass *)nm->itemOld.lParam;
+
+    if (info != nullptr) {
+        if (info->m_type == ASSET_TYPE_TEXTURE) {
+            Ref_Ptr_Release(info->m_texture);
+        }
+
+        delete info;
+    }
+
+    GetTreeCtrl().SetItem(nm->itemOld.hItem, TVIF_PARAM, nullptr, 0, 0, 0, 0, 0);
+    *pResult = 0;
+}
+
+void CDataTreeView::OnDblClk(NMHDR *pNMHDR, LRESULT *pResult)
+{
+    *pResult = 0;
+}
+
+void CDataTreeView::AddRenderObjects()
+{
+    SetRedraw(FALSE);
+    RenderObjIterator *iter = W3DAssetManager::Get_Instance()->Create_Render_Obj_Iterator();
+
+    if (iter != nullptr) {
+        iter->First();
+
+        while (!iter->Is_Done()) {

```

Comment:
Maybe can make this a compact for loop `for (iter->First(), !iter->Is_Done(), iter->Next)`

Reply:
Fixed.